test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,258 @@
+    def camel_case_to_snake_case(camel_case_str):
+        return re.sub(r'(?<!^)(?=[A-Z])', '_', camel_case_str).lower()
+
+    if 'cpp_options_args' in test_params_dict:
+        # Expected format for `cpp_options_args`: `F::FunctionalFuncOptions(...)`
+        # Example output: `binary_cross_entropy`
+        return camel_case_to_snake_case(
+            test_params_dict['cpp_options_args'].split('(')[0].replace('F::', '').replace('FuncOptions', ''))
+    elif 'cpp_function_call' in test_params_dict:
+        # Expected format for `cpp_function_call`: `F::functional_name(...)`
+        # Example output: `binary_cross_entropy`
+        return test_params_dict['cpp_function_call'].split('(')[0].replace('F::', '')
+    else:
+        raise RuntimeError(
+            "`cpp_options_args` or `cpp_function_call` entry must be present in test params dict:\n{}".format(
+                pprint.pformat(test_params_dict)))
+
+def compute_cpp_function_call(test_params_dict, arg_dict, functional_name):
+    if 'cpp_function_call' in test_params_dict:
+        return test_params_dict['cpp_function_call']
+    elif 'cpp_options_args' in test_params_dict:
+        cpp_forward_args_symbols = [arg_name for arg_name, _ in 
+                                    arg_dict['input'] + arg_dict['target'] + arg_dict['extra_args']]
+        return 'F::{}({}, {})'.format(
+            functional_name, ", ".join(cpp_forward_args_symbols), test_params_dict['cpp_options_args'])
+    else:
+        raise RuntimeError(
+            "`cpp_options_args` or `cpp_function_call` entry must be present in test params dict:\n{}".format(
+                pprint.pformat(test_params_dict)))
+
+def process_test_params_for_functional(test_params_dict, device, test_instance_class):
+    test_instance = test_instance_class(**test_params_dict)
+    functional_name = compute_functional_name(test_params_dict)
+    assert test_instance.get_name().startswith('test_')
+    # Example output: `BCELoss_no_reduce_cuda`
+    functional_variant_name = test_instance.get_name()[5:] + (('_' + device) if device != 'cpu' else '')
+    arg_dict = compute_arg_dict(test_params_dict, test_instance)
+
+    return TorchNNFunctionalTestParams(
+        functional_name=functional_name,
+        functional_variant_name=functional_variant_name,
+        test_instance=test_instance,
+        cpp_function_call=compute_cpp_function_call(test_params_dict, arg_dict, functional_name),
+        arg_dict=arg_dict,
+        has_parity=test_params_dict.get('has_parity', True),
+        device=device,
+        cpp_tmp_folder=tempfile.mkdtemp(),
+    )
+
+def test_torch_nn_functional_variant(unit_test_class, test_params):
+    test_forward(unit_test_class, test_params)
+
+def add_torch_nn_functional_impl_parity_tests(
+        parity_table, unit_test_class, test_params_dicts, test_instance_class, devices):
+    if not hasattr(unit_test_class, 'functional_test_params_map'):
+        unit_test_class.functional_test_params_map = {}
```

Comment:
Can we just add an empty dict to the unit_test_class instead of checking every time?